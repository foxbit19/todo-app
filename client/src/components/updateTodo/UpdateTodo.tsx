import { Button, Dialog, DialogActions, DialogContent, DialogContentText, DialogTitle, TextField } from '@mui/material'
import React, { useEffect, useState } from 'react'
import Item from '../../models/item'
import GenericDialog from '../dialog/GenericDialog'

interface Props {
    open: boolean
    item: Item,
    onUpdateClick?: (updated: Item) => void
    onClose?: () => void
}

const UpdateTodo = (props: Props) => {
    const [inputValue, setInputValue] = useState<string>(props.item.description)

    const handleChange = (event: any) => {
        setInputValue(event.target.value)
    }

    const handleUpdateClick = () => {
        if (props.onUpdateClick) {
            props.onUpdateClick(new Item(props.item.id, inputValue, props.item.order))
        }
    }

    useEffect(() => {
        if (props.item) {
            setInputValue(props.item.description)
        }
    }, [props.item])


    return (
        <GenericDialog open={props.open}
            title='Item update'
            text={'Edit the todo description in order to update it.'}
            button={<Button data-testid='update_button' variant='contained' onClick={handleUpdateClick}>Update</Button>}
            onClose={props.onClose}>
            <TextField autoFocus required fullWidth margin='dense' label='Description' data-testid='description' value={inputValue} onChange={handleChange} variant='standard' />
        </GenericDialog>
    )
}

export default UpdateTodo