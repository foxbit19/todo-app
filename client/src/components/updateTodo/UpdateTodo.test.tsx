import React from 'react';
import { render, screen } from '@testing-library/react';
import Item from '../../models/item';
import sinon from 'sinon';
import UpdateTodo from './UpdateTodo';
import userEvent from '@testing-library/user-event';

describe('update todo', () => {
    let item: Item

    beforeEach(() => {
        item = new Item(1, 'this is my item', 2)
    })

    const getInputElement = (): HTMLInputElement => {
        return screen.getByTestId('description')
    }

    const updateButtonClick = () => {
        const buttonElement: HTMLButtonElement = screen.getByTestId('update_button')
        buttonElement.click()
    }

    test('it matches snapshot', () => {
        const view = render(<UpdateTodo open={true} item={item} />);
        expect(view).toMatchSnapshot();
    });

    test('it shows the "item update" text', () => {
        render(<UpdateTodo open={true} item={item} />)

        expect(screen.getByText('Item update')).toBeInTheDocument()
    })

    test('it shows the description of the provided item into an input box', () => {
        render(<UpdateTodo open={true} item={item} />)

        expect(getInputElement().value).toEqual(item.description)
    })

    test('it launch an event when the update button is pressed', () => {
        const updateHandler = sinon.spy();

        render(<UpdateTodo open={true} item={item} onUpdateClick={updateHandler} />)
        updateButtonClick()

        expect(updateHandler.called).toBeTruthy()
    })

    test('it launch an event with the modified item when the update button is pressed', () => {
        const updateHandler = sinon.spy();

        render(<UpdateTodo open={true} item={item} onUpdateClick={updateHandler} />)

        const newDescription = 'remember to test everything'

        userEvent.clear(getInputElement())
        userEvent.type(getInputElement(), newDescription)
        updateButtonClick()

        expect(updateHandler.calledWith({
            id: item.id,
            description: newDescription,
            order: item.order
        })).toBeTruthy()
    })
})