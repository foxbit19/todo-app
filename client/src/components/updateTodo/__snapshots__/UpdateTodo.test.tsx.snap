// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update todo it matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby=":r0:"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthLg css-1fu2e3p-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
            id=":r0:"
          >
            Item update
          </h2>
          <div
            class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
          >
            <p
              class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiDialogContentText-root css-qfso29-MuiTypography-root-MuiDialogContentText-root"
            >
              Edit the todo description in order to update it.
            </p>
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
              data-testid="description"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for=":r1:"
                id=":r1:-label"
              >
                Description
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                  â€‰
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  id=":r1:"
                  required=""
                  type="text"
                  value="this is my item"
                />
              </div>
            </div>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Close
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="update_button"
              tabindex="0"
              type="button"
            >
              Update
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
